STORM-2281: Running Multiple Kafka Spouts (Trident) Throws Illegal State Exception

hmcl opened the pull request:
To review this PR please review only the commit that has STORM-2281. It is currently sitting on top of the multiple commits addressing STORM-2225, waiting for them to be squashed in and eventually merged.

@harshach can you please review. Thanks.

HeartSaVioR commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitter.java:
```diff
@@ -161,31 +187,84 @@ private long seek(TopicPartition tp, KafkaTridentSpoutBatchMetadata<K, V> lastBa
-        final Set<TopicPartition> pausedTopicPartitions  = new HashSet<>(kafkaConsumer.assignment());
-        LOG.debug("Currently assigned topic partitions [{}]", pausedTopicPartitions);
+        final Set<TopicPartition> pausedTopicPartitions = new HashSet<>(kafkaConsumer.assignment());
+        LOG.trace("Currently assigned topic partitions [{}]", pausedTopicPartitions);
```
Do we need to change this to trace?

hmcl replied:
Done

harshach commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitter.java:
```diff
@@ -47,70 +33,111 @@
+     * Creates instance of this class with default 500 millisecond refresh subscription timer
+     */
+    public KafkaTridentSpoutEmitter(KafkaTridentSpoutManager<K,V> kafkaManager, TopologyContext topologyContext) {
+        this(kafkaManager, topologyContext, new Timer(500,
```
do we want make the timer here configurable instead of 500

HeartSaVioR replied:
This is for backward compatible. Another constructor which receives timer is available.

harshach commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitter.java:
```diff
@@ -47,70 +33,111 @@
+        } else {
+            try {
+                // pause other topic partitions to only poll from current topic partition
+                pausedTopicPartitions = pauseTopicPartitions(currBatchTp);
```
how this will work in case of multiple executors running the spout. It looks like we are making sure there is only executor/kafka-consumer that can poll the topic partitions. It looks like the intent here is to keep data from 1 partition per batch.

hmcl replied:
I think it will work fine. Each executor will have one instance of `KakfkaConsumer`. The org.apache.storm.trident.spout.OpaquePartitionedTridentSpoutExecutor.Emitter class has some logic that emits one batch per topic-partition. I believe the same happens in the Kafka Spout with the old consumer. In case I am misunderstanding, can you please elaborate a bit more. Thanks.

harshach commented on pom.xml:
```diff
@@ -267,7 +267,7 @@
 
         <!-- kafka version used by new storm-kafka-client spout code -->
-        <storm.kafka.client.version>0.10.0.0</storm.kafka.client.version>
+        <storm.kafka.client.version>0.10.1.1</storm.kafka.client.version>
```
Lets keep the version to 0.10.0.1 here. Its better to support lowest version as much as possible. As of now old clients can connect to new brokers but not other way round.

hmcl replied:
Done

harshach commented:
@hmcl overall looks good to me. Left few questions. @hmcl has some good points around the topic names.

hmcl commented on examples/storm-kafka-examples/src/main/java/org/apache/storm/kafka/trident/TridentKafkaWordCount.java:
```diff
@@ -95,7 +95,8 @@ public static void main(String[] args) throws Exception {
                 localSubmitter.submit(consTpName, tpConf, TridentKafkaConsumerTopology.newTopology(localSubmitter.getDrpc(),
-                        new TransactionalTridentKafkaSpout(newTridentKafkaConfig(zkBrokerUrl[0]))));
+                        new OpaqueTridentKafkaSpout(newTridentKafkaConfig(zkBrokerUrl[0]))));
+//                        new TransactionalTridentKafkaSpout(newTridentKafkaConfig(zkBrokerUrl[0]))));
```
I meant to leave it such that it's pretty trivial, and also brings awareness, of using this test to test both Opaque and Transactional Trident. I think it's an overdo to write code supporting both (e.g. with a CLI option), that's why I left this as an easy way to do so. I can put a comment saying why this code is there. If you still think it should be removed, I can do it. Let me know.

HeartSaVioR replied:
Example topology doesn't really need to provide cover features exhaustively. Users still can see TransactionalTridentKafkaSpout is available, even not looking at this class. If users can't find, IMHO it's a documentation issue.

hmcl replied:
Done

hmcl commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitter.java:
```diff
@@ -47,70 +33,111 @@
-            // pause other topic partitions to only poll from current topic partition
-            pausedTopicPartitions = pauseTopicPartitions(topicPartition);
+        if (assignments == null || !assignments.contains(currBatchPartition.getTopicPartition())) {
+            LOG.debug("SKIPPING processing batch: [transaction = {}], [currBatchPartition = {}], [lastBatchMetadata = {}], " +
```
I agree with WARN. I believe the the code can only get here if there is a rebalance while the tasks are being assigned. This logic will guarantee that we retry.

hmcl commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutBatchMetadata.java:
```diff
@@ -74,8 +74,8 @@ public TopicPartition getTopicPartition() {
     public String toString() {
-        return "KafkaTridentSpoutBatchMetadata{" +
-                "topicPartition=" + topicPartition +
+        return super.toString() +
```
All this is doing is that the method `toString()` instead of printing `KafkaTridentSpoutBatchMetadata` it will print `org.apache.storm.kafka.spout.trident.KafkaTridentSpoutBatchMetadata@6b74f986`. This was useful while I was debugging and writing the code because the code is multi threaded for multiple executors, and I needed to see which object instance (e.g `@6b74f986`) was being passed around. I can revert it back, but for me there was benefit in having it like this.

HeartSaVioR replied:
Thanks for the details. I'm OK to leave this.

hmcl commented on examples/storm-kafka-client-examples/src/main/java/org/apache/storm/kafka/trident/TridentKafkaClientWordCountNamedTopics.java:
```diff
@@ -107,8 +113,12 @@ protected void run(String[] args) throws AlreadyAliveException, InvalidTopologyE
+                    }
 
                     // print
                     localSubmitter.printResults(15, 1, TimeUnit.SECONDS);
```
Done

hmcl commented on examples/storm-kafka-client-examples/src/main/java/org/apache/storm/kafka/trident/TridentKafkaClientWordCountNamedTopics.java:
```diff
@@ -107,8 +113,12 @@ protected void run(String[] args) throws AlreadyAliveException, InvalidTopologyE
                     localSubmitter.submit(topic2Tp, tpConf, KafkaProducerTopology.newTopology(brokerUrl, topic2));
                     // Consumer
+                    try {
                     localSubmitter.submit(consTpName, tpConf, TridentKafkaConsumerTopology.newTopology(
```
Done

hmcl commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitter.java:
```diff
@@ -196,8 +275,8 @@ public void close() {
     public String toString() {
-        return "KafkaTridentSpoutEmitter{" +
-                ", kafkaManager=" + kafkaManager +
+        return super.toString() +
```
Same comment as above.

hmcl commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutManager.java:
```diff
@@ -107,9 +90,25 @@ Fields getFields() {
     public String toString() {
-        return "KafkaTridentSpoutManager{" +
-                "kafkaConsumer=" + kafkaConsumer +
+        return super.toString() +
```
Same comment as above.

hmcl commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutOpaque.java:
```diff
@@ -79,8 +72,8 @@ public Fields getOutputFields() {
     public String toString() {
-        return "KafkaTridentSpoutOpaque{" +
-                "kafkaManager=" + kafkaManager +
+        return super.toString() +
```
Same comment as above.

hmcl commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutOpaqueCoordinator.java:
```diff
@@ -57,8 +57,8 @@ public void close() {
     public String toString() {
-        return "KafkaTridentSpoutOpaqueCoordinator{" +
-                "kafkaManager=" + kafkaManager +
+        return super.toString() +
```
Same comment as above.

hmcl commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutTopicPartition.java:
```diff
@@ -43,7 +43,7 @@ public TopicPartition getTopicPartition() {
     @Override
     public String getId() {
-        return topicPartition.topic() + "/" + topicPartition.partition();
+        return topicPartition.topic() + "_" + topicPartition.partition();
```
so which one do you think it should be? This is new code that I wrote, so it should be a matter of choice. Before this change I had **/** but I felt that there was no need to create a new zookeeper level, hence I changed it to **_**. **_** is what exists in the old kafka consumer code. Please take a look at method `org.apache.storm.kafka.Partition#getId()` and let me know which one we should go with. Currently the code works using either **/** or **_**

HeartSaVioR replied:
Yes it will work if no one is willing to extract the information via parsing the return value. OK to leave as it is, but if possible I'd like to avoid the possible confusing case. 

If you're worrying about creating redundant znode because of '/', how about choosing the character which is not legal char for kafka topic but legal char for znode name, for example, `@`?

hmcl commented on examples/storm-kafka-examples/src/main/java/org/apache/storm/kafka/trident/LocalSubmitter.java:
```diff
@@ -43,10 +43,15 @@ public static LocalSubmitter newInstance() {
         conf.setMaxSpoutPending(20);
-        conf.setMaxTaskParallelism(1);
-        conf.setNumWorkers(1);
+//        conf.setMaxTaskParallelism(2);
```
Done

hmcl commented:
@harshach I should add this to master branch as well, right ?

hmcl commented:
@harshach @HeartSaVioR I believe I have addressed all your review comments. I temporarily left one commit just with the code review diff. I will squash it once we agree the patch is good to go. Can you please do one last pass. Thanks.

hmcl commented:
@harshach @HeartSaVioR can you please take a look. Thanks.

harshach commented:
+1.

harshach commented:
@hmcl yes we need one for master as well.

HeartSaVioR commented:
@hmcl Thanks for the update. +1 from me. There's a conflict while cherry-picking into master, so we might need to have following PR for master.

hmcl commented:
@harshach @HeartSaVioR thanks for the review. I have squashed the commits. It should be good to go.

HeartSaVioR commented:
+1 again.
